### PR TITLE
Refactor and small optimization for dynamic content html (part 2/2)

### LINKED ISSUE #9993: Larger amount of images leads to performance problems
Issue:
I have 270 images in the images folder. For each image the size is analyzed:
```
// Helper/FileManager.php
if ($size = @getimagesize($filepath)) {
    $files[] = [
        'src'    => $fullurl,
        'width'  => $size[0],
        'type'   => 'image',
        'height' => $size[1],
    ];
} 
```
With 270 images this leads to load times >20s. And this not only on the loading of the asset manager in Grapesjs, but also on the regular Mautic pages like Dashboard, Contacts, etc.

### LINKED ISSUE #9997: Cant do space or delete or backspace for text in buttons
Cant do space or delete or backspace for text in buttons

Steps to recreate:

Add button to email/page.

Type text in button and hit space bar then type text.

Observer space is not done.

Same goes for backspace and delete.

https://drive.google.com/file/d/1bnKWwe62puqXzrSQRYJDB2FW1Z4gM4iw/view?usp=sharing

Could it be because of this code:

https://github.com/Webmecanik/plugin-grapesjs-builder/blob/6fd88c362333084dd12732f03f6790113753db84/Assets/js/builder.js#L118-L125

### LINKED ISSUE #10000: Page Builder: CSS not loaded
While editing a page in the Page Builder the CSS is not loaded. It is fine once you save, and open the page with the Preview or Live URL.

image

Initial debugging indicates that `getAsssetUrl` is not working:
image

EDIT: 31.8
I further investigated this. And I realized that absolute URLs are also not loaded in the page builder. So it should **not** be about `getAsssetUrl`:
image

### LINKED ISSUE #10142: GrapesJS email builder wipes out HTML head tags
## Bug Description
Everything that is not in body section is removed from the email HTML when using GrapesJS builder.

| Q   | A
| --- | ---
| Mautic version | 3.3.3 and 4.0.0-beta
| PHP version | 7.4
| Browser | Any

### Steps to reproduce
1. Create email and select template that has HTML head section e.g. Blank
2. Open the builder
3. Close builder and save email
4. Open preview or send example
5. Head tags, custom html declaration or anything except body is replaced with default values

### LINKED ISSUE #10246: CKE Editor is loaded "outside" the container
## Bug Description
When I drop a new Dynamic Content Block on the grapesjs canvas the CKEditor is loaded bigger than the popup:
image

Maybe this is only a problem with the PR https://github.com/mautic/mautic/pull/10114

| Q   | A
| --- | ---
| Mautic version | 
| PHP version | 
| Browser | 

### Steps to reproduce
1. Maybe: Load PR https://github.com/mautic/mautic/pull/10114
2. edit a new email
3. add dynamic content block in grapesjs
 

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/99860493-cke-editor-is-loaded-outside-the-container?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by adiux**
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features"
| Bug fix?                               | yes/no
| New feature?                           | no
| Deprecations?                          | yes
| BC breaks?                             | could be
| Automated tests included?              | no (only JS changes, currently no tests infrastructure available)
| Related user documentation PR URL      | no
| Related developer documentation PR URL | there are many comments directly in the code
| Issue(s) addressed                     | working towards enabling dynamic content for MJML fixes #10142 #9993

#### Description:
Background: There are two repositories for Mautic Grapesjs: Plugin and Preset. It was decided that we move all the general Mautic functionality in the Preset (core functionality, library). The plugin itself is only the implementation of that. It should contain only a minimum of code. Only the code necessary for wiring Grapesjs into Mautic (similar to a controller). This PR takes the first step in that direction.
- moving code from plugin to preset
- using classes and separate files for better code maintainability
- importing external libraries via npm and JS: `import`

#### No more known Issue => fixed
~There is a bug with the dependency on CKEditor. So the dynamic content (textareas) can not be edited. Until this is resolved, testing is only possible for general functionality.~

fixes: https://github.com/mautic/mautic/issues/10142
fixes: https://github.com/mautic/mautic/issues/9997
fixes: https://github.com/mautic/mautic/issues/10246
fixes: https://github.com/mautic/mautic/issues/10000

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Create a new dynamic content block
3. Send the email and verify that the correct block is displayed.

**Please test the steps above both with the new GrapesJS builder as well as the legacy builder**.

### CODE REVIEW

**Comment by alexhammerschmied** [comment]
While loading up the PR in Mautibox it says "patch could not be applied" and gives an error. :(

**Comment by RCheesley** [comment]
@alexhammerschmied please try testing locally as Mautibox has some issues with Mautic 4 PRs that we haven't yet been able to resolve.

[Here](https://www.youtube.com/watch?v=dOQX66xY5gs&t=96s) is a vid where I explained how to set up DDEV for our last sprint - it's a bit rambly as it was unplanned so will do a better one in the future!

**Comment by patrykgruszka** [comment]
After applying this PR and clearing cache the legacy builder is loading instead of GrapesJS (plugin is enabled). Am I missing something?

**Comment by RCheesley** [comment]
Have you enabled the builder in the plugins section? https://docs.mautic.org/builders has some instructions in case that is helpful.

**Comment by patrykgruszka** [comment]
Yes, I did. I have this running on the features branch.

**Comment by RCheesley** [comment]
@adiux after applying this PR I am seeing errors in the console which are not there with features:

emails:199 GET https://mautic4.ddev.site/plugins/GrapesJsBuilderBundle/Assets/library/js/dist/builder.js?v8deada2a net::ERR_ABORTED 500
emails:189 GET https://mautic4.ddev.site/plugins/GrapesJsBuilderBundle/Assets/library/js/dist/builder.css net::ERR_TOO_MANY_REDIRECTS

Also getting the same as @patrykgruszka - the legacy builder is being loaded rather than GrapesJS, with new emails and with existing ones that were created in GrapesJS!

So something is quite broken it would seem.

**Comment by RCheesley** [comment]
Still seem to be getting the same issue with the latest commits :(

**Comment by adiux** [comment]
It works for me locally, maybe we can do a screen share this afternoon?

**Comment by tejasync** [comment]
Hi @adiux,
I have enabled the GrapeJS plugin.
I'm trying this in a incognito window, I get this error in the console:
Screenshot 2021-06-17 at 17 08 07

I looked up the file explorer, couldn't find the files either:
Screenshot 2021-06-17 at 17 09 17

**Comment by adiux** [comment]
@ts-navghane I just committed the dist/ files. Thanks for testing!

**Comment by RCheesley** [comment]
<s>@adiux we have a conflict to resolve here after merging the PR to enable the builder by default - would you be able to take a look please?</s> it was a tiny fix so I have resolved it!

**Review by RCheesley — CHANGES_REQUESTED** [review]
PR works as expected with the exception of the editor on dynamic content which we are aware of. Otherwise was able to create the emails and send them.

I found an issue on landing pages, however.

The old builder UI is showing, and the following error is given in the console:  
﻿
```
builder.service.js:137 Uncaught TypeError: l.getAssetManagerConf is not a function
    at l.value (builder.service.js:137)
    at l.value (builder.service.js:97)
    at Object.s [as launchBuilder] (builder.js:30)
    at HTMLButtonElement.onclick (pages:1)
    at Object.trigger (2.jquery.js?va81aa353:8255)
    at HTMLButtonElement.<anonymous> (2.jquery.js?va81aa353:8327)
    at Function.each (2.jquery.js?va81aa353:354)
    at jQuery.fn.init.each (2.jquery.js?va81aa353:189)
    at jQuery.fn.init.trigger (2.jquery.js?va81aa353:8326)
    at jQuery.fn.init.jQuery.fn.<computed> [as click] (2.jquery.js?va81aa353:10212)
```

Builder is definitely active so it seems like there may be something that is broken here.

**Comment by RCheesley** [comment]
Findings from the latest test:

1. Landing page errors are gone, works great :tada: 
        - DWC block is missing
       - Couple of errors in the console which don't seem to be causing a problem:    

```
6CanvasView.js:205 Uncaught TypeError: Cannot read property 'frameElement' of null
    at r.getFrameOffset (CanvasView.js:205)
    at r.getElementPos (CanvasView.js:230)
    at r.getElementPos (SelectComponent.js:716)
    at r.updateGlobalPos (SelectComponent.js:202)
    at r.onFrameUpdated (SelectComponent.js:146)
    at Jt (_executeBound.js:8)
    at bind.js:10
    at restArguments.js:16
```

`img/video2.webm:1 GET https://mautic4.ddev.site/s/pages/edit/img/video2.webm net::ERR_ABORTED 404`

Just dumped a video block in and didn't set anything up so that might explain this error.

2. Emails seem to have a couple of issues - tested with Brienz:
         - <s>Dynamic Content block is missing (maybe still working on it?)</s> (this PR doesn't add the block for MJML themes - that is to come in the future)
         - Code editor view is not working - nothing in the console but it does not open (works great in landing page)
         - <s>A new block appeared, looks like 'raw' - had thought this was a code mode block but it drops in some images. They look nice in the editor but are wonky in the preview:</s> Created a separate issue as not related to this PR

**Comment by adiux** [comment]
https://github.com/mautic/mautic/issues/10240

**Comment by RCheesley** [comment]
I tested this PR and the dynamic content is working in the editor now - also code mode works in HTML emails and in MJML now.

Code mode is *not* working with the landing page builder however (using the blank theme). Nothing happens when I click the code mode button.

Unfortunately when the email is sent to a segment the dynamic content variations that are set in the email are not being used - everybody is getting the default content. It seems that the variants are not getting saved with the email - when I edit the email after saving they are all gone.

Recording here: https://youtu.be/A58w9fxvCxc

**Comment by adiux** [comment]
Great that code mode works for you in HTML now! I'm not sure what you are saying with MJML. I have it working for Email MJML: 
image

Or which code mode are you talking about? The current one? The one that I think we should remove so we don't have to maintain two different code modes? https://github.com/mautic/mautic/issues/9988

**Comment by RCheesley** [comment]
Sorry it looks like I had some issue with assets getting cached on an old localhost environment (I had one spun up using mautic.ddev.site for a video I was doing which I later deleted and was then testing in mautic4.ddev.site and for some reason the builder was still looking for mautic.ddev.site)

**Comment by dennisameling** [comment]
Yesterday and today we had two issues with Dynamic Content in the new builder:
- Both Froala and CKEditor would be loaded at the same time, which led to weird behavior (https://github.com/mautic/mautic/issues/10246)
- After removing Froala and just activating CKEditor, there was a bug where **existing** Dynamic Content tabs could not be edited, but **newly created** Dynamic Content tabs worked well ("Variation 2" in the video below is a new tab):

https://user-images.githubusercontent.com/17739158/126669513-c410fd36-f9b2-4c82-8f2c-4a35dea87eaf.mp4

The second problem occurred because the Dynamic Content `textarea` elements were converted into CKEditor instances **on page load** already, so before the Email Builder was even open. That caused some sort of conflict because of which those textareas weren't editable, and performance-wise it's not ideal to do all of that on page load already (imagine the email having 10+ Dynamic Content instances...). This behavior was present in the legacy builder already. 

Both issues have been fixed in https://github.com/mautic/mautic/pull/10114/commits/6b9405d7acad27280c5036bc2a62fbecccbb27e8 by doing the following:
- Do not convert the Dynamic Content `textarea` elements into CKEditor instances on page load anymore (applies to non-legacy builders only)
- Added new function `Mautic.setDynamicContentEditors()` which can be called by builders like the GrapesJsBuilder, as soon as the editing modal opens. This turns said textareas into CKEditor instances, which are now correctly editable.
- It's possible to add new tabs (filters) to Dynamic Content, but we didn't have a way to turn those new tabs into CKEditor instances as well. The new `Mautic.dynamicContentAddNewItemListener(callback)` and `Mautic.dynamicContentAddNewFilterListener(callback)` functions can now be leveraged by builders to do anything they want with textareas of newly created Dynamic Content slots or filters. In the case of the GrapesJsBuilder, this means - once again - turning those into CKEditor instances as well
- When the modal is closed, we destroy the CKEditor instances so that next time users open the same modal, it creates a fresh instance, preventing the issues of not being able to edit that data again.

The result is that users can now edit and create Dynamic Content items with the CKEditor:

https://user-images.githubusercontent.com/17739158/126671236-416b7988-dc91-4d5a-bc73-afa293aa4246.mp4

Note that it's not yet possible to save the contents of Dynamic Content items, @adiux is working on that 😊

**Comment by dennisameling** [comment]
Looks like we broke something in this PR in the legacy builder. The default value isn't saved (but the additional filters are):

https://user-images.githubusercontent.com/17739158/126673984-342f2e96-3c09-4626-a546-2471c8def7f4.mp4

Something to look into once we get further into processing Dynamic Content data in the new builder! 🤷🏼

**Comment by adiux** [comment]
I think the way CKE is handled currently is differently to Froala. 

I can see the content while the editor is open. 
Screen Shot 2021-07-27 at 15 56 44

when the editor gets destroyed the values are gone.
Screen Shot 2021-07-27 at 16 19 46


Maybe related to the editor being destroyed?

**Comment by dennisameling** [comment]
@adiux I just tried with the Froala editor using Mautic's `features` branch, and the content is also not saved (despite me clicking "Save" in the bottom left corner).

image

So the problem doesn't seem to be editor-related.

> Maybe related to the editor being destroyed?

I remember seeing something in the code that was related to "getting Froala's contents" so the frontend could actually save it in the HTML code. It was in one of the JS files of Mautic itself. I don't remember which one it was though 😅 

I'll try to have a look at this either tomorrow or on Friday, but last time this whole Dynamic Content thing proved to be super tricky so can't promise anything...

**Comment by adiux** [comment]
@dennisameling  There was a fix necessary to save Froala. It was working before the change to CKE. I really think this is related to the CKE being destroyed. I debugged it and the values get removed from the dome during the for loop that destroys the editor. 

first they are here
Screen Shot 2021-08-06 at 12 30 37

then they get removed from the dom:
https://user-images.githubusercontent.com/13075514/128498805-5ce5f4e2-acf1-41cb-bd3c-03788c7ba38c.mp4

The tricky part to understand is that the dynamic content values (content and filter) are stored in the textarea - or the editor. At `updateDcStoreItem()` I have to copy the whole editor from Grapesjs back to the Mautic <form> so it can get persisted when the email is saved.

**Comment by dennisameling** [comment]
> The tricky part to understand is that the dynamic content values (content and filter) are stored in the textarea - or the editor. At updateDcStoreItem() I have to copy the whole editor from Grapesjs back to the Mautic
> 
> so it can get persisted when the email is saved.

@adiux First I couldn't find `updateDcStoreItem()` (only in the minified JS), but now see you've introduced it in a PR to the preset repo that's still open: https://github.com/mautic/grapesjs-preset-mautic/pull/6/files - I wasn't aware of that PR, good to know :)

As you mentioned, we destroy the CKEditor indeed, but that [should simply write the contents back to the textarea](https://docs-old.ckeditor.com/ckeditor_api/symbols/CKEDITOR.editor.html#destroy). Apparently it doesn't looking at your screencast, but I think that has to do with GrapesJS having its own DOM somehow.

Let me have a closer look at your preset PR (https://github.com/mautic/grapesjs-preset-mautic/pull/6/files) tomorrow, as I think it's now simply a matter of getting the CKEditor contents (something like `CKEDITOR.instances.editor1.getData()`) prior to destroying it, like you mentioned you did for Froala in the past. Think we're almost there then (famous last words 😉)

**UPDATE**: I now see that I did `CKEDITOR.instances[name].destroy(true)` instead of `CKEDITOR.instances[name].destroy(false)` - the `true` refers to [`noUpdate`](https://docs-old.ckeditor.com/ckeditor_api/symbols/CKEDITOR.editor.html#destroy) in CKEditor - setting that to `false` should write the data back into the textarea, but I'll need to test that hypothesis. To be continued 👍🏼

**Comment by dennisameling** [comment]
I'd like to have a final look at the code changes we made in this PR before we start testing extensively. Will do that either later today or tomorrow. CC @RCheesley

File: app/bundles/EmailBundle/Assets/js/email.js [hunk 1]
```diff
@@ -330,12 +333,20 @@ Mautic.createNewDynamicContentItem = function(jQueryVariant) {
...
+            // Set custom buttons with separator between them.
+            toolbarSticky: false,
+            toolbarButtons: ['undo', 'redo', '|', 'bold', 'italic', 'underline', 'paragraphFormat', 'fontFamily', 'fontSize', 'color', 'align', 'formatOL', 'formatUL', 'quote', 'clearFormatting', 'token', 'insertLink', 'insertImage', 'insertTable', 'html', 'fullscreen'],
+            heightMin: 100
+        }));
+    }
+
+    if (Mautic.internalDynamicContentItemCreateListeners) {
+        Mautic.internalDynamicContentItemCreateListeners.forEach(function(callback) {
+            callback(textarea);
```
Comment (dennisameling) [inline app/bundles/EmailBundle/Assets/js/email.js:347]: I'd like to move Dynamic Content logic to its own JS file after we've merged this PR, to allow for better compatibility with external builders in the future, as well as an consistent API for those (creating listeners etc.). Curious to hear your thoughts on that @adiux

**Review by dennisameling — DISMISSED** [review]
This is looking good to me! Created a Dynamic Content slot both in the legacy builder and in the GrapesJS builder, and when I send an email through a campaign, the logic works correctly:

Email created in the GrapesJS builder:

image

Email created in the legacy builder:

image

I did notice the following things though:
- The spacing of the Dynamic Content slot is slightly different in the GrapesJS rendered email compared to the legacy rendered email (see screenshots above, a padding of 10px is automatically added by GrapesJS).
- The Mautic styling is missing in the GrapesJS builder. This was already reported in https://github.com/mautic/grapesjs-preset-mautic/issues/16
- When sending an email to a contact manually, the Dynamic Content slot is not rendered (a first name token does get rendered though). I checked and this bug was already present, so unrelated to this PR.

https://user-images.githubusercontent.com/17739158/129448122-263ca953-d8da-4c77-a935-0222273dad10.mp4

The issues above are non-blocking IMO and can be solved at a later stage. 

@adiux I see there was some code commented out in the `Demo/helloWorld/helloWorld.js` file, was that intentional?

**Comment by RCheesley** [comment]
I've taken the following steps:

1. Install Mautic from current features branch
2. Check out this PR
3. Open GrapesJS and create a new email **(segment)**, drag Dynamic Content onto the canvas

```
2libraries.js?v0c1970f3:366 Uncaught TypeError: Cannot call a class as a function
    at l (builder.js:31)
    at HTMLDocument.i (builder.js:31)
    at mightThrow (libraries.js:355)
    at process (libraries.js:357)
```
^ This appears as soon as I drag the dynamic content over onto the canvas.

4. Create a default and a variant, and save
5. Exit the builder

```
builder.js?v0c1970f3:31 Uncaught TypeError: t.deleteDynamicContentItem is not a function
    at r.<anonymous> (builder.js?v0c1970f3:31)
    at m (builder.js?v0c1970f3:3)
    at v (builder.js?v0c1970f3:3)
    at d (builder.js?v0c1970f3:3)
    at r.c.trigger (builder.js?v0c1970f3:3)
    at removeChildren (builder.js?v0c1970f3:3)
    at builder.js?v0c1970f3:3
    at Module.ke (builder.js?v0c1970f3:3)
    at r.forEach (builder.js?v0c1970f3:3)
    at removeChildren (builder.js?v0c1970f3:3)
```
^ This appears as soon as I exit the builder.

The email is sent but there is no dynamic content at all being displayed - default or otherwise :(

**Comment by RCheesley** [comment]
@adiux since your last commit I am now getting this error as soon as I open the email builder:

```
Uncaught TypeError: Cannot read property 'set' of null
    at r.<anonymous> (grapesjs-preset-newsletter.min.js:15)
    at m (backbone.js:369)
    at v (backbone.js:356)
    at d (backbone.js:155)
    at r.l.trigger (backbone.js:346)
    at EditorView.js:17
```

When I save the dynamic content, the variations are no longer available and are therefore not being sent out in the email. I am no longer getting the errors I was seeing before, however!

**Comment by adiux** [comment]
`Cannot read property 'set' of null` is not related to Mautic. As mentioned here: https://github.com/mautic/mautic/pull/10349#issuecomment-903529683

**Comment by adiux** [comment]
It looks like you have somehow gotten an old version of the code. I just tested on my branch with segement and template emails. With mjml and with html mode. 

Thank you for testing! Im not sure why this is... Maybe you can checkout the branch directly?

The [RTE PR](https://github.com/mautic/mautic/pull/10349) is based on this PR.

**Review by RCheesley — APPROVED** [review]
Yassss!  All working now - dynamic content going to the right people and the builder is working as expected!

screenshot-mautic4 ddev site_8026-2021 08 23-13_42_39
screenshot-mautic4 ddev site_8026-2021 08 23-13_42_59
screenshot-mautic4 ddev site_8026-2021 08 23-13_43_20


🚀

**Pull request merged**